### ISSUE TITLE
`useless_let_if_seq` issue?

### ISSUE THREAD

**Issue by frederikhors** [rust-lang/rust-clippy#15739]
I have this code:

```rs
let mut size: usize = 0;
let mut page: u64 = 1;

if let Some(pagination) = pagination {
    size = pagination.size().try_into().unwrap();
    page = pagination.page();
}
```

and clippy is saying:

```
warning: `if _ { .. } else { .. }` is an expression
239 | /     let mut page: u64 = 1;
240 | |
241 | |     if let Some(pagination) = pagination {
242 | |         size = pagination.size().try_into().unwrap();
243 | |         page = pagination.page();
244 | |     }
    | |_____^ help: it is more idiomatic to write: `let <mut> page = if let Some(pagination) = pagination { ..; pagination.page() } else { 1 };`
    |
    = note: you might not need `mut` at all
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_let_if_seq
    = note: `-W clippy::useless-let-if-seq` implied by `-W clippy::nursery`
    = help: to override `-W clippy::nursery` add `#[allow(clippy::useless_let_if_seq)]`
```

So I think I should change it to:

```rs
let size = input
    .pagination
    .as_ref()
    .map_or(0, |pagination| pagination.size().try_into().unwrap());

let page = input.pagination.as_ref().map_or(1, Input::page);
```

Am I right?

Isn't the check on the `.map_or()` duplicated now? Is this a performance regression?

**Comment by nickdrozd** [comment]
`useless_let_if_seq` output is pretty bad. For one thing, it deliberately outputs invalid code (https://github.com/rust-lang/rust-clippy/issues/1081). And it doesn't do much analysis.

But the intent of the lint is to flag variables that are defined and then immediately mutated. Idea is to define variables all in one go.

In the case here, I think you could simplify to:

```rust
let (size, page): (usize, u64) = if let Some(pagination) = pagination {
    (pagination.size().try_into().unwrap(), pagination.page())
} else {
    (0, 1)
};
```

Avoids mutation, checks `pagination` only once.

**Comment by frederikhors** [comment]
Thanks @nickdrozd, I already fixed it like this. But I'm wondering if should be doing something with this issue, maybe not.
